From 33db51679d278c29c7d6dd51541246ecbfeff20b Mon Sep 17 00:00:00 2001
From: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
Date: Fri, 30 May 2014 14:02:12 +0100
Subject: [PATCH 272/351] ASoC: arizona: Add DVFS handling for sample rates 2
 and 3

This change implements DVFS for sample rate domains 2 and 3

Change-Id: I8fc6c2ee00633cc4935168bd04f6e2449f992aeb
Signed-off-by: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
---
 sound/soc/codecs/arizona.c |   41 +++++++++++++++++++++++++++++++++++++++++
 sound/soc/codecs/arizona.h |   11 +++++++++++
 sound/soc/codecs/florida.c |    4 ++--
 sound/soc/codecs/wm5102.c  |    4 ++--
 4 files changed, 56 insertions(+), 4 deletions(-)

diff --git a/sound/soc/codecs/arizona.c b/sound/soc/codecs/arizona.c
index 0f8fdd8..2c32b7a 100644
--- a/sound/soc/codecs/arizona.c
+++ b/sound/soc/codecs/arizona.c
@@ -1595,6 +1595,47 @@ static int arizona_sr_vals[] = {
 	512000,
 };
 
+int arizona_put_sample_rate_enum(struct snd_kcontrol *kcontrol,
+				 struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_codec *codec = snd_kcontrol_chip(kcontrol);
+	struct arizona *arizona = dev_get_drvdata(codec->dev->parent);
+	struct soc_enum *e = (struct soc_enum *)kcontrol->private_value;
+	unsigned int val;
+	unsigned int flag;
+	int ret;
+
+	ret = snd_soc_put_value_enum_double(kcontrol, ucontrol);
+	if (ret == 0)
+		return 0;	/* register value wasn't changed */
+
+	val = e->values[ucontrol->value.enumerated.item[0]];
+
+	switch (e->reg) {
+	case ARIZONA_SAMPLE_RATE_2:
+		flag = ARIZONA_DVFS_SR2_RQ;
+		break;
+
+	case ARIZONA_SAMPLE_RATE_3:
+		flag = ARIZONA_DVFS_SR3_RQ;
+		break;
+
+	default:
+		return ret;
+	}
+
+	if (arizona_sr_vals[val] >= 88200) {
+		ret = arizona_dvfs_up(arizona, flag);
+		if (ret != 0)
+			dev_err(codec->dev, "Failed to raise DVFS %d\n", ret);
+	} else {
+		ret = arizona_dvfs_down(arizona, flag);
+	}
+
+	return ret;
+}
+EXPORT_SYMBOL_GPL(arizona_put_sample_rate_enum);
+
 static int arizona_startup(struct snd_pcm_substream *substream,
 			   struct snd_soc_dai *dai)
 {
diff --git a/sound/soc/codecs/arizona.h b/sound/soc/codecs/arizona.h
index dbac523..0e9485c 100644
--- a/sound/soc/codecs/arizona.h
+++ b/sound/soc/codecs/arizona.h
@@ -183,6 +183,14 @@ extern int arizona_mixer_values[ARIZONA_NUM_MIXER_INPUTS];
 	ARIZONA_MIXER_ROUTES(name " Preloader", name "L"), \
 	ARIZONA_MIXER_ROUTES(name " Preloader", name "R")
 
+#define ARIZONA_SAMPLE_RATE_CONTROL(name, domain) \
+	SOC_VALUE_ENUM(name, arizona_sample_rate[(domain) - 2])
+
+#define ARIZONA_SAMPLE_RATE_CONTROL_DVFS(name, domain)        \
+	SOC_ENUM_EXT(name, arizona_sample_rate[(domain) - 2], \
+			snd_soc_get_value_enum_double,        \
+			arizona_put_sample_rate_enum)
+
 #define ARIZONA_RATE_ENUM_SIZE 4
 #define ARIZONA_SAMPLE_RATE_ENUM_SIZE 14
 #define ARIZONA_OUT_RATE_ENUM_SIZE 3
@@ -231,6 +239,9 @@ extern int arizona_anc_ev(struct snd_soc_dapm_widget *w,
 			  struct snd_kcontrol *kcontrol,
 			  int event);
 
+extern int arizona_put_sample_rate_enum(struct snd_kcontrol *kcontrol,
+					struct snd_ctl_elem_value *ucontrol);
+
 extern int arizona_set_sysclk(struct snd_soc_codec *codec, int clk_id,
 			      int source, unsigned int freq, int dir);
 
diff --git a/sound/soc/codecs/florida.c b/sound/soc/codecs/florida.c
index c09e027..78f0e76 100644
--- a/sound/soc/codecs/florida.c
+++ b/sound/soc/codecs/florida.c
@@ -440,8 +440,8 @@ SOC_ENUM("LHPF2 Mode", arizona_lhpf2_mode),
 SOC_ENUM("LHPF3 Mode", arizona_lhpf3_mode),
 SOC_ENUM("LHPF4 Mode", arizona_lhpf4_mode),
 
-SOC_VALUE_ENUM("Sample Rate 2", arizona_sample_rate[0]),
-SOC_VALUE_ENUM("Sample Rate 3", arizona_sample_rate[1]),
+ARIZONA_SAMPLE_RATE_CONTROL("Sample Rate 2", 2),
+ARIZONA_SAMPLE_RATE_CONTROL("Sample Rate 3", 3),
 
 SOC_VALUE_ENUM("ISRC1 FSL", arizona_isrc_fsl[0]),
 SOC_VALUE_ENUM("ISRC2 FSL", arizona_isrc_fsl[1]),
diff --git a/sound/soc/codecs/wm5102.c b/sound/soc/codecs/wm5102.c
index 7e54596..9f5db37 100644
--- a/sound/soc/codecs/wm5102.c
+++ b/sound/soc/codecs/wm5102.c
@@ -874,8 +874,8 @@ SOC_ENUM("LHPF2 Mode", arizona_lhpf2_mode),
 SOC_ENUM("LHPF3 Mode", arizona_lhpf3_mode),
 SOC_ENUM("LHPF4 Mode", arizona_lhpf4_mode),
 
-SOC_VALUE_ENUM("Sample Rate 2", arizona_sample_rate[0]),
-SOC_VALUE_ENUM("Sample Rate 3", arizona_sample_rate[1]),
+ARIZONA_SAMPLE_RATE_CONTROL_DVFS("Sample Rate 2", 2),
+ARIZONA_SAMPLE_RATE_CONTROL_DVFS("Sample Rate 3", 3),
 
 SOC_VALUE_ENUM("ISRC1 FSL", arizona_isrc_fsl[0]),
 SOC_VALUE_ENUM("ISRC2 FSL", arizona_isrc_fsl[1]),
-- 
1.7.9.5

